File: scene2.c

#include <render_oo.h>
#include <scenedef.h>
#include <models.h>
#include <render_ray.h>

// scene.background = (.8, .8, .8)
// TODO: make scene.background work (I don't have image->clear)
int main() {
    init_scene();
    struct Camera* cam = get_camera();
    struct Scene* scene = get_scene();
    cam->set_perspective(cam, 60, 1.3333, 5);

    struct Box b = make_box((struct Point3F32){-3, -2, -10}, (struct Vec3){2, 2, 2}, make_pixel(1, 0, 0));

    scene->add(scene, &(struct Shape){BOX, .shape.b = b});

    struct Image img_wf = make_image(640, 480);
    render_wireframe(scene, &img_wf);
    img_wf.save(&img_wf, "ppm/scene2-wf.ppm");

    struct Image img_sig = make_image(640, 480);
    render_signature(scene, &img_sig);
    img_sig.save(&img_sig, "ppm/scene2-sig.ppm");
    free_image(&img_sig);

    struct Image img_rt = make_image(640, 480);
    raytrace(scene, &img_rt);
    img_rt.save(&img_rt, "ppm/scene2-rt.ppm");

    return 0;
}

File: src/scene2.zig
const std = @import("std");
const scenedef = @import("3d/scenedef.zig");
const render_ray = @import("3d/render_ray.zig");
const models = @import("3d/models.zig");
const materials = @import("3d/materials.zig");
const gfx_types = @import("3d/gfx_types.zig");
const image = @import("3d/image.zig");
const raytrace = @import("3d/render_ray.zig");

pub fn main() !void {
    var debug_alloc = std.heap.DebugAllocator(.{}).init;
    const alloc = debug_alloc.allocator();

    try scenedef.init_scene(alloc);
    defer scenedef.deinit_scene(alloc);

    var cam = scenedef.get_camera();
    const scene = scenedef.get_scene();
    cam.set_perspective(60, 1.3333, 5);

    var b = models.Shape{
        .box = models.Box.init(.{ .vals = .{ -3, -2, -10 } }, .{ .vals = .{ 2, 2, 2 } }, materials.make_material(gfx_types.Pixel{ .r = 1, .g = 0, .b = 0 })),
    };

    try scene.add(&b);

    const cwd = try std.fs.cwd().realpathAlloc(alloc, ".");
    defer alloc.free(cwd);
    const img_path = try std.fs.path.join(alloc, &.{ cwd, "ppm/scene2-rt.ppm" });
    defer alloc.free(img_path);

    var img = try image.Image.init(alloc, 640, 480);
    defer img.deinit(alloc);
    raytrace.raytrace(scene, &img);
    try img.save(img_path);
}
